pages/blog/posts/rxdb-case-study.md

```diff
@@ -0,0 +1,223 @@
+        }
+    },
+    additionalProperties: false,
+    required: ['firstName', 'lastName', 'id'],
+    // RxDB specific fields:
+    primaryKey: 'id',
+    version: 0,
+    keyCompression: true,
+    indexes: [
+        'firstName', // single-field index
+        ['firstName', 'lastName'] // compound index
+    ],
+    encrypted: ['secret']
+}
+```
+
+## Restrictions
+
+Although RxDB aims to remain fully spec compliant with JSON Schema, it does impose some extra restrictions:
+
+The schema requires top-level field names to match the regex `^[a-zA-Z][a-zA-Z0-9_]*[a-zA-Z0-9]$`. This rule ensures that property names are valid JavaScript identifiers.
+
+At the top level of the schema, `additionalProperties` must be set to `false`. This prevents silently introducing new fields that could potentially clash with built-in RxDocument methods (like `.toJSON()`). By making you list out each property, RxDB ensures you can't unknowingly overwrite or conflict with standard methods.
+
+RxDB does not allow `$ref` to other files or external schema fragments or `$dynamicRef`. The goal is to keep schema loading fast and self-contained. If a schema needs to be composed of multiple parts, you must combine them at build time or otherwise ensure they are merged before passing them to RxDB at runtime. This approach also prevents any network calls or asynchronous fetches that could slow down your application's startup.
+
+
+## Inferring Document Types with TypeScript
+
+JSON Schema is not just for validation and structural guarantees, it can also help generate or infer TypeScript types. In many projects, developers rely on tools like [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript) to produce `.d.ts` files or interface definitions from schema files at build time. However, that requires a separate build step, which slows down your workflow.
+
+To improve developer experience, RxDB offers a [built-in way](https://rxdb.info/tutorials/typescript.html) to infer the document's TypeScript type from the schema during runtime. This is helpful because you get immediate feedback in your IDE: As soon as you update the schema, TypeScript picks up the changes. You'll see type errors in your code if you attempt to use fields that are no longer valid or if you forget to include newly required fields.
+
+Below is an example on how to interfere the TypeScript type of a document from its JSON-schema:
+
+```ts
+import {
+    toTypedRxJsonSchema,
+    ExtractDocumentTypeFromTypedRxJsonSchema,
+    RxJsonSchema,
+    RxCollection
+} from 'rxdb';
+
+const heroSchemaLiteral = {
+    title: 'hero schema',
+    description: 'describes a human being',
+    version: 0,
+    keyCompression: true,
+    primaryKey: 'passportId',
+    type: 'object',
+    properties: {
+        passportId: {
+            type: 'string',
+            maxLength: 100
+        },
+        /* ...other fields... */
+    },
+    required: ['firstName', 'lastName', 'passportId'],
```

Comment:
It's completely irrelevant to the article but a pet peeve in database structures is having first and last name fields (rather than just a "name") field (c.f. https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/)